onnxruntime/core/optimizer/qdq_transformer/qdq_util.cc

```diff
@@ -215,4 +215,141 @@ bool MatchDQNode(const Node& node) {
+    return false;
+  }
+
+  // Support scalar float scale only for now. Need to extend to other float types if needed.
+  Initializer scale_initializer(*scale_tensor_proto, graph_viewer.ModelPath());
+  if (scale_initializer.dims().size() != 0 || scale_initializer.data_type() != ONNX_NAMESPACE::TensorProto::FLOAT) {
+    return false;
+  }
+  scale = *scale_initializer.data<float>();
+
+  const ONNX_NAMESPACE::TensorProto* zp_tensor_proto =
+      graph_viewer.GetConstantInitializer(q_input_defs[2]->Name(), true);
+  if (!zp_tensor_proto) {
+    return false;
+  }
+
+  Initializer zp_initializer(*zp_tensor_proto, graph_viewer.ModelPath());
+  if (zp_initializer.dims().size() != 0) {
+    return false;
+  }
+
+  data_type = zp_initializer.data_type();
+  switch (data_type) {
+    case ONNX_NAMESPACE::TensorProto::INT8:
+      zp = static_cast<int32_t>(*zp_initializer.data<int8_t>());
+      break;
+    case ONNX_NAMESPACE::TensorProto::UINT8:
+      zp = static_cast<int32_t>(*zp_initializer.data<uint8_t>());
+      break;
+    case ONNX_NAMESPACE::TensorProto::INT16:
+      zp = static_cast<int32_t>(*zp_initializer.data<int16_t>());
+      break;
+    case ONNX_NAMESPACE::TensorProto::UINT16:
+      zp = static_cast<int32_t>(*zp_initializer.data<uint16_t>());
+      break;
+    default:
+      return false;
+  }
+
+  return true;
+}
+
+bool CanRemoveRelu(const GraphViewer& graph_viewer, const Node& q_node) {
+  float scale = 0.0f;
+  int32_t zp = 0;
+  int32_t data_type = 0;
+  if (!GetQSalarScaleZp(graph_viewer, q_node, scale, zp, data_type)) {
+    return false;
+  }
+
+  int32_t data_type_min = 0;
+  int32_t data_type_max = 0;
+  if (!GetDataTypeMinMax(data_type, data_type_min, data_type_max)) {
+    return false;
+  }
+
+  // Relu can be removed if the zero-point is set to the smallest quantized value.
+  return zp == data_type_min;
+}
+
+bool CanRemoveClip(const GraphViewer& graph_viewer, const Node& clip_node, const Node& q_node) {
+  float scale = 0.0f;
+  int32_t zp = 0;
+  int32_t data_type = 0;
+  if (!GetQSalarScaleZp(graph_viewer, q_node, scale, zp, data_type)) {
+    return false;
+  }
+
+  float min = 0.0f;
+  float max = 0.0f;
+  if (!optimizer_utils::GetClipConstantMinMax(graph_viewer.GetGraph(), clip_node, min, max)) {
+    return false;
+  }
+
+  int32_t q_clip_min = static_cast<int32_t>(::rint(min / scale)) + zp;
+  int32_t q_clip_max = static_cast<int32_t>(::rint(max / scale)) + zp;
+
+  int32_t data_type_min = 0;
+  int32_t data_type_max = 0;
+  if (!GetDataTypeMinMax(data_type, data_type_min, data_type_max)) {
+    return false;
+  }
+
+  // The Clip can be removed if its range entirely overlaps the quantization range.
+  // QClip range:    [------------------]
+  // Quant range:      [-------------]
+  return q_clip_min <= data_type_min && q_clip_max >= data_type_max;
+}
+
+}  // namespace
+
+bool CanFuseActivationQ(const GraphViewer& graph_viewer, const Node& activation_node, const Node& q_node) {
```

Comment:
To Scott's previous point, perhaps this can be renamed to something more like `IsClipMadeRedundantByQ` (both relu and clip are clipping operations). Also, I realize that you're probably using the "fusion" terminology because that is what I originally used in the original QNN EP code, so that's on me :).